src/api-binder.ts

```diff
@@ -48,390 +53,470 @@ interface DeviceTag {
-				}),
-			).timeout(timeout)) as Device;
-		} catch (e) {
-			return null;
-		}
+function getStateDiff(): DeviceStatus {
+	const lastReportedLocal = lastReportedState.local;
+	const lastReportedDependent = lastReportedState.dependent;
+	if (lastReportedLocal == null || lastReportedDependent == null) {
+		throw new InternalInconsistencyError(
+			`No local or dependent component of lastReportedLocal in ApiBinder.getStateDiff: ${JSON.stringify(
+				lastReportedState,
+			)}`,
+		);
 	}
 
-	public async patchDevice(id: number, updatedFields: Dictionary<unknown>) {
-		const conf = await config.getMany([
-			'unmanaged',
-			'provisioned',
-			'apiTimeout',
-		]);
+	const diff = {
+		local: _(stateForReport.local)
+			.omitBy((val, key: keyof DeviceStatus['local']) =>
+				_.isEqual(lastReportedLocal[key], val),
+			)
+			.omit(INTERNAL_STATE_KEYS)
+			.value(),
+		dependent: _(stateForReport.dependent)
+			.omitBy((val, key: keyof DeviceStatus['dependent']) =>
+				_.isEqual(lastReportedDependent[key], val),
+			)
+			.omit(INTERNAL_STATE_KEYS)
+			.value(),
+	};
+
+	return _.omitBy(diff, _.isEmpty);
+}
 
-		if (conf.unmanaged) {
-			throw new Error('Cannot update device in unmanaged mode');
-		}
+async function sendReportPatch(
+	stateDiff: DeviceStatus,
+	conf: { apiEndpoint: string; uuid: string; localMode: boolean },
+) {
+	if (balenaApi == null) {
+		throw new InternalInconsistencyError(
+			'Attempt to send report patch without an API client',
+		);
+	}
 
-		if (!conf.provisioned) {
-			throw new Error('DEvice must be provisioned to update a device');
+	let body = stateDiff;
+	if (conf.localMode) {
+		body = stripDeviceStateInLocalMode(stateDiff);
+		// In local mode, check if it still makes sense to send any updates after data strip.
+		if (_.isEmpty(body.local)) {
+			// Nothing to send.
+			return;
 		}
+	}
 
-		if (this.balenaApi == null) {
-			throw new InternalInconsistencyError(
-				'Attempt to patch device without an API client',
-			);
-		}
+	const endpoint = url.resolve(
+		conf.apiEndpoint,
+		`/device/v2/${conf.uuid}/state`,
+	);
+
+	const requestParams = _.extend(
+		{
+			method: 'PATCH',
+			url: endpoint,
+			body,
+		},
+		balenaApi.passthrough,
+	);
+
+	await balenaApi._request(requestParams);
+}
 
-		return Bluebird.resolve(
-			this.balenaApi.patch({
-				resource: 'device',
-				id,
-				body: updatedFields,
-			}),
-		).timeout(conf.apiTimeout);
+// Returns an object that contains only status fields relevant for the local mode.
+// It basically removes information about applications state.
+export function stripDeviceStateInLocalMode(state: DeviceStatus): DeviceStatus {
+	return {
+		local: _.cloneDeep(
+			_.omit(state.local, 'apps', 'is_on__commit', 'logs_channel'),
+		),
+	};
+}
+
+const report = _.throttle(async () => {
+	const conf = await config.getMany([
+		'deviceId',
+		'apiTimeout',
+		'apiEndpoint',
+		'uuid',
+		'localMode',
+	]);
+
+	const stateDiff = getStateDiff();
+	if (_.size(stateDiff) === 0) {
```

Comment:
It's worth doing this before the config fetch, as with the return, we can avoid a whole config.getMany